File: 2236-MaximumTwinSumOfALinkedList/2236-MaximumTwinSumOfALinkedList.py
# Last updated: 6/29/2026, 10:21:10 AM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        # reverse half link
        # TC for bruteforce => O(n * n);
        # optimal appproach -> reverse the 1st half ll & itr 1st & 2nd half

        fast = head
        slow = head
        prev = None
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            head.next = prev
            prev = head
            head = slow

        # head & slow hai 
        res = 0
        while slow:
            curr = prev.val + slow.val
            res = max(res, curr)
            prev = prev.next
            slow = slow.next
        return res

        

File: 2236-MaximumTwinSumOfALinkedList/2236-MaximumTwinSumOfALinkedList.cpp
// Last updated: 6/29/2026, 10:21:16 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;
        
        ListNode* prev = nullptr;
        ListNode* curr = head;
        ListNode* nxt = head->next;

        while(fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
            nxt = nxt->next;
        }
        int res = 0;
        while(slow != nullptr) {
            res = max(res, slow->val + prev->val);
           prev = prev->next; slow = slow->next;
        }
        return res;
    }
};